factor: improve large integer factorization

mattsu2020 opened the pull request:
## Summary

- improve factorization of integers larger than `u128`
- use small-prime trial division, Miller-Rabin primality testing, Pollard p-1, and a Brent-style Pollard rho implementation
- keep the existing `num-prime` fast paths for `u64` and `u128` inputs
- preserve the existing input parsing and error-reporting behavior
- add a regression test for a large composite number

## Motivation

The previous `BigUint` path delegated factorization to `num-prime` and could fail to completely factor large inputs. This change adds bounded fallback algorithms for large integers while retaining the existing optimized paths for smaller values.

This addresses the performance problem reported in [Ubuntu bug 2131212](https://bugs.launchpad.net/ubuntu/+source/rust-coreutils/+bug/2131212) and is related to #10262.

## Implementation notes

Factorization first removes small prime factors, then checks probable primality with Miller-Rabin. Composite values are processed with Pollard p-1 followed by a Brent-style Pollard rho search. The search uses bounded retries and iteration counts so difficult inputs do not run indefinitely. If complete factorization is not achieved, `factor` keeps the partial result and returns a non-zero exit status.

## Validation

- `cargo fmt --all -- --check`
- `cargo clippy -p uu_factor --all-targets --tests --benches -- -D warnings`
- `cargo test -p uu_factor`
- `cargo test --test tests --features factor test_factor::test_large_regression_number -- --exact`

mattsu2020 commented on src/uu/factor/src/factor.rs:
```diff
@@ -123,6 +133,356 @@ fn write_result_u128(
+            return Some(g);
+        }
+
+        // a^(M) を段階的に構成（指数を 2^k で伸ばす近似）
```
Sorry. Done

sylvestre commented:
could you please run hyperfine with the three programs? gnu, without the patch and with the patch
and share the full results here? thanks :)

sylvestre commented on src/uu/factor/src/factor.rs:
```diff
@@ -123,6 +133,356 @@ fn write_result_u128(
+    fn lcg_next(x: &mut u128) {
+        *x = x
+            .wrapping_mul(6364136223846793005)
+            .wrapping_add(1442695040888963407);
```
same

sylvestre commented on src/uu/factor/src/factor.rs:
```diff
@@ -123,6 +133,356 @@ fn write_result_u128(
+    // Search parameters: choose bounds based on bit length.
+    // Avoid overly large limits; when exhausted, treat as failure to find a factor.
+    let max_tries: u64 = 16;
+    let max_iter: u64 = (bits * bits).clamp(10_000, 200_000);
```
why these values ?

mattsu2020 commented on src/uu/factor/src/factor.rs:
```diff
@@ -123,6 +133,356 @@ fn write_result_u128(
+        return true;
+    }
+    // even check: candidate % 2 == 0
+    if (candidate & BigUint::from_u32(1).unwrap()).is_zero() {
```
Done

mattsu2020 commented on src/uu/factor/src/factor.rs:
```diff
@@ -123,6 +133,356 @@ fn write_result_u128(
+    let mut odd_component = candidate - &one;
+    let mut power_of_two = 0u32;
+    // while odd_component is even
+    while (&odd_component & BigUint::from_u32(1).unwrap()).is_zero() {
```
Done

mattsu2020 commented on src/uu/factor/src/factor.rs:
```diff
@@ -123,6 +133,356 @@ fn write_result_u128(
+    let max_tries: u64 = 16;
+    let max_iter: u64 = (bits * bits).clamp(10_000, 200_000);
+
+    let mut seed: u128 = 0x9e3779b97f4a7c15;
```
Done

sylvestre commented:
many jobs are failing

mattsu2020 committed:
factor: improve large integer factorization

Use Miller-Rabin primality testing with Pollard p-1 and Brent-style Pollard rho for values larger than u128. Preserve the existing input parsing behavior and add a regression test for a large composite.

sylvestre commented:
Sorry not sure I understand why we should merge this PR

mattsu2020 commented:
> Sorry not sure I understand why we should merge this PR

I updated the title and PR description.
To improve compatibility with GNU coreutils, we are adding support for large numbers.
